### PR TITLE
CLEO V2.0.0.7

### PR DESCRIPTION
**Pull request by XMDS**
fix https://github.com/cleolibrary/III.VC.CLEO/issues/16

### CODE REVIEW

File: source/III.VC.CLEO/Fxt.cpp [hunk 1]
```diff
@@ -24,6 +24,40 @@ CustomTextEntry::~CustomTextEntry()
 		delete m_pText;
 }
 
+static uint8_t get_utf8_bytes(uint8_t utf8)
+{
+	for (uint8_t i = 0; i < 6; i++) {
```
Comment (MiranDMC) [inline source/III.VC.CLEO/Fxt.cpp:29]: This project is using "opening bracket in new line" style. Better keep it consistent.

File: source/III.VC.CLEO/Fxt.cpp [hunk 2]
```diff
@@ -24,6 +24,40 @@ CustomTextEntry::~CustomTextEntry()
...
+		if ((utf8 & (0x80 >> i)) == 0) {
+			return i == 0 ? 1 : i;
+		}
+	}
+	return 1;
+}
+
+uint16_t CustomText::convert_utf8_to_utf16(uint8_t* utf8, uint16_t utf8_len, uint16_t* utf16, uint16_t utf16_len)
```
Comment (MiranDMC) [inline source/III.VC.CLEO/Fxt.cpp:37]: Use char\wchar_t for characters, int\unsigned int for counts.

File: source/III.VC.CLEO/Fxt.cpp [hunk 3]
```diff
@@ -24,6 +24,40 @@ CustomTextEntry::~CustomTextEntry()
...
+{
+	uint16_t length = 0;
+	uint8_t bytes = 0;
+
+	for (uint16_t i = 0; i < utf8_len && length < utf16_len; i++, length++) {
+		bytes = get_utf8_bytes(utf8[i]);
+
+		if (bytes > 1) {
+			utf16[length] = utf8[i] & (0xFF >> (bytes + 1));
```
Comment (MiranDMC) [inline source/III.VC.CLEO/Fxt.cpp:46]: Many magic in single line. How about creating variable named "mask", then using in. Comments would be nice too.

File: source/III.VC.CLEO/Fxt.cpp [hunk 4]
```diff
@@ -24,6 +24,40 @@ CustomTextEntry::~CustomTextEntry()
...
+			for (uint8_t j = 1; j < bytes; j++) {
+				i++;
+				utf16[length] <<= 6;
+				utf16[length] += utf8[i] & 0x3F;
```
Comment (MiranDMC) [inline source/III.VC.CLEO/Fxt.cpp:50]: Magic numbers? 0x3F comes out of blue without any explanation.

File: source/III.VC.CLEO/Fxt.cpp [hunk 5]
```diff
@@ -24,6 +24,40 @@ CustomTextEntry::~CustomTextEntry()
...
+	uint16_t length = 0;
+	uint8_t bytes = 0;
+
+	for (uint16_t i = 0; i < utf8_len && length < utf16_len; i++, length++) {
+		bytes = get_utf8_bytes(utf8[i]);
+
+		if (bytes > 1) {
+			utf16[length] = utf8[i] & (0xFF >> (bytes + 1));
+			for (uint8_t j = 1; j < bytes; j++) {
+				i++;
```
Comment (MiranDMC) [inline source/III.VC.CLEO/Fxt.cpp:48]: Can now **i** address out of data buffer?

**Review by MiranDMC — CHANGES_REQUESTED** [review]

File: source/III.VC.CLEO/Fxt.cpp [hunk 6]
```diff
@@ -9,10 +9,10 @@ CustomTextEntry *CustomText::pCustomTextList;
 CustomTextEntry::CustomTextEntry(char *key, char *text)
 {
     size_t len = strlen(text);
-	this->m_pText = new wchar_t[len + 1];
-	for(size_t i = 0; i < len; i++)
-		this->m_pText[i] =  (unsigned char)text[i];
-	this->m_pText[len] = 0;
+	uint16_t* u16_text = new uint16_t[len + 1];
```
[the remark this answers is not in the document]
Comment (XMDS) [reply source/III.VC.CLEO/Fxt.cpp:12]: In fact, it is better to use the uint16_t type, first of all it is more intuitive to tell us that it is 2 bytes (although the cleo library runs on win, but I have planned to port it to the mobile platform, the wchar_t type is not 2 bytes in some systems  ), another reason is that my experience in Chinese games tells me that you should never use the wchar_t type to save Chinese characters.

[the remark this answers is not in the document]
Comment (XMDS) [reply source/III.VC.CLEO/Fxt.cpp:12]: In fact, it is better to use the uint16_t type, first of all it is more intuitive to tell us that it is 2 bytes (although the cleo library runs on win, but I have planned to port it to the mobile platform, the wchar_t type is not 2 bytes in some systems  ), another reason is that my experience in Chinese games tells me that you should never use the wchar_t type to save Chinese characters.

File: source/III.VC.CLEO/Fxt.cpp [hunk 7]
```diff
@@ -9,10 +9,10 @@ CustomTextEntry *CustomText::pCustomTextList;
...
+	memset(u16_text, 0, len + 1);
```
[the remark this answers is not in the document]
Comment (XMDS) [reply source/III.VC.CLEO/Fxt.cpp:13]: Yes.  I forgot, now it only has half the memory written with 0's.
Forgive me I prefer c style

File: source/III.VC.CLEO/Fxt.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (MiranDMC) [reply source/III.VC.CLEO/Fxt.cpp:12]: Perhaps then
`typedef uint16_t wchar` is most elegant solution.

**Comment by MiranDMC** [comment]
Please rebase on current master branch. We added some of similar changes already, so this PR will become simpler.

**Pull request closed without merging**